extensions/data-transfer/portability-data-transfer-flickr/src/main/java/org/datatransferproject/datatransfer/flickr/media/FlickrMediaImporter.java

```diff
@@ -0,0 +1,332 @@
+            auth = FlickrUtils.getAuth(authData, flickr);
+        } catch (FlickrException e) {
+            return new ImportResult(e);
+        }
+        RequestContext.getRequestContext().setAuth(auth);
+
+        Preconditions.checkArgument(
+                data.getAlbums() != null || data.getPhotos() != null || data.getVideos() != null,
+                "Error: There is no data to import");
+
+        if (data.getAlbums() != null) {
+            storeAlbums(jobId, data.getAlbums());
+        }
+
+        if (data.getPhotos() != null) {
+            for (PhotoModel photo : data.getPhotos()) {
+                try {
+                    importSinglePhoto(idempotentExecutor, jobId, photo);
+                } catch (FlickrException e) {
+                    if (e.getMessage().contains("Upload limit reached")) {
+                        throw new DestinationMemoryFullException("Flickr destination memory reached", e);
+                    } else if (e.getMessage().contains("Photo already in set")) {
+                        // This can happen if we got a server error on our end, but the request went through.
+                        // When our retry strategy kicked in the request was complete and the photo already
+                        // uploaded
+                        continue;
+                    }
+                    throw new IOException(e);
+                }
+            }
+        }
+
+        if (data.getVideos() != null) {
+            for (VideoModel video : data.getVideos()) {
+                try {
+                    importSingleVideo(idempotentExecutor, jobId, video);
+                } catch (FlickrException e) {
+                    if (e.getMessage().contains("Upload limit reached")) {
+                        throw new DestinationMemoryFullException("Flickr destination memory reached", e);
+                    } else if (e.getMessage().contains("Video already in set")) {
+                        // This can happen if we got a server error on our end, but the request went through.
+                        // When our retry strategy kicked in the request was complete and the photo already
+                        // uploaded
+                        continue;
+                    }
+                    throw new IOException(e);
+                }
+            }
+        }
+
+        return new ImportResult(ImportResult.ResultType.OK);
+    }
+
+    // Store any album data in the cache because Flickr only allows you to create an album with a
+    // photo in it, so we have to wait for the first photo to create the album
+    private void storeAlbums(UUID jobId, Collection<MediaAlbum> albums) throws IOException {
+        for (MediaAlbum album : albums) {
+            jobStore.create(
+                    jobId,
+                    ORIGINAL_ALBUM_PREFIX + album.getId(),
+                    new FlickrTempPhotoData(album.getName(), album.getDescription()));
+        }
+    }
+
+    private void importSinglePhoto(
```

Comment:
LMK if you want to delete some of this duplicated code; I'd be happy to help factor it out so we don't have two drifting codepaths (eg see microsoft version where we call the [same method two times and the rest of the call stack is generic to models](https://sourcegraph.com/github.com/jzacsh/data-transfer-project@e00d33478bb91f983c29d7183966704ac6d4f8e4/-/blob/extensions/data-transfer/portability-data-transfer-microsoft/src/main/java/org/datatransferproject/transfer/microsoft/media/MicrosoftMediaImporter.java?L117))

Reply:
That sounds good, I added a TODO to come back and do that